tests/replication/_base.py

```diff
@@ -339,7 +340,7 @@ def make_worker_hs(
         # `_handle_http_replication_attempt` like we do with the master HS.
         instance_name = worker_hs.get_instance_name()
         instance_loc = worker_hs.config.worker.instance_map.get(instance_name)
-        if instance_loc:
+        if instance_loc and isinstance(instance_loc, InstanceTCPLocationConfig):
```

Comment:
Seems like we could throw an error when we see `InstanceUNIXLocationConfig` here so we don't run into other unrelated errors masking this root cause

Reply:
No, but it need to assert that a `host` attribute was present. Since Twisted's unit tests take care of testing that Unix sockets function, there appeared to be no value in adding that infrastructure to the testing setup in Synapse.

Reply:
Your reply is a little unclear.

It's fine not to test unix sockets but if someone does try to use unix sockets in some tests iin the future, I want to make sure we fail with a clear error that unix sockets are either not supported yet or they just work fine.

Reply:
That is an interesting idea. I actually hadn't considered that someone might actually *try* and do that. I will explore where to add a good "Don't do that" error message 😄

Reply:
That was an interesting rabbit hole I fell down, looking into this more. Apparently, the `MemoryReactor` that is used as the base for the unit tests *does* support Unix sockets, after a fashion. It is essentially a `Mock`-like Fake system(more or less, more likely less). 

I don't mind adding an error message into *this* spot, keeping in mind this is for the `instance_map` settings, so the subset of tests it would apply to is narrow and would require a complicated `instance_map` and `listener`/`worker_listeners` setup.

I think for the "non-`instance_map`" sockets, the most universal option is to override the `connectUNIX()` and `listenUNIX` methods of the `MemoryReactor`/`ThreadedMemoryReactor` so if they are ever called it will raise an Exception. When testing this hypothesis, the only test I could get to fail(after I added such a config to the `default_config` as the most universal spot) was `tests.http.test_site.SynapseRequestTestCase.test_large_request`. It seems to be the only one that actually goes through all the spots to actually *use* any kind of socket(AFAICT). Everything else just seems to be Fake. Having said all that, I may just be completely wrong about how I went about it, as the unit test system is really complex.

As an alternative: add big warnings to the docs about not bothering to write any tests for Unix sockets perspective. Although I'm skeptical about how much such a warning would be read, all things considered.

Reply:
<details><summary>Code I tested with(added to test_sharded_event_persister), to hookup instance_map</summary>

```python
    def default_config(self) -> dict:
        conf = super().default_config()
        conf["stream_writers"] = {"events": ["worker1", "worker2"]}
        conf["instance_map"] = {
            "main": {"host": "testserv", "port": 8765},
            "worker1": {"host": "testserv", "port": 1001},
            "worker2": {"path": "/var/run/worker2.sock"},
        }
        conf["listeners"] = [
            {
                "path": "/var/run/mainpublic.sock",
                "type": "http",
                "resources": [
                    {
                        "names": ["client", "federation"]
                    }
                ]
            }
        ]
        conf["worker_listeners"] = [
            {
                "path": "/var/run/worker2.sock",
                "type": "http",
                "resources": [
                    {
                        "names": ["client", "federation"]
                    }
                ]
            }
        ]

        return conf

```
</details>

Reply:
Added them in [ee55d2a](https://github.com/matrix-org/synapse/pull/15708/commits/ee55d2a275651cfa657af4bbfabbd1fc53fa63dd) so you can judge if that's ok.